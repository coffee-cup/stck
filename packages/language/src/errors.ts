import { capitalize } from "lodash";
import { Range } from "./types";

export class EvalError extends Error {
  constructor(message: string, range: Range) {
    super(message);
    (this as any).location = range;
  }
}

export const typeError = (
  operator: string,
  expected: string,
  received: any,
): string =>
  `${capitalize(operator)} expected: ${expected}, received: ${received}`;
