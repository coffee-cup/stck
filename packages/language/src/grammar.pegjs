{
  const makeNode = (type, value) => ({
    type,
    ...value,
    ...location()
  })
}

start = program

program 
  = _ body:body _
  { return makeNode("program", { body }) }

expr
  = loop
  / function
  / call
  / pushpop
  / operator
  / comment

pushpop "pushpop"
  = left:(literal/identifier)? ">" right:identifier?
  { return makeNode("pushpop", { left, right  } ) }  
  
operator "operator"
  = left:identifier op:OP right:(literal/identifier)?
  { return makeNode("operator", { left, op, right }) }
  
OP = "+" / "-" / "*" / "/" / "?"

function "function"
 = "{" _ name:fnname _ body:body _ "}"
 { return makeNode("function", { name, body }) }

loop "loop"
  = "(" _ id:identifier _ body:body _ ")"
  { return makeNode("loop", { stack: id.value, body }) }
  
body
 = exprs:(_ e:expr { return e })*
 { return exprs }
  
call "call"
  = name:fnname
  { return makeNode("call", { name }) }
  
fnname 
  = ":" id:identifier
  { return id.value }

identifier "identifier" 
  = [a-zA-Z]+ 
  { return makeNode("identifier", {value: text()}) }
  
literal
  = number
  / string
  
string "string"
  = '"' chars:stringChars* '"' 
  { return makeNode("literal", { value: chars.join("") }) }
  
stringChars
  = !('"' / "\\") . { return text() }
  / "\\" seq:"\"" { return seq }

number "number"
  = minus? int frac? 
  { return makeNode("literal", {value: parseFloat(text())}) }

comment "comment"
  = "#" p:([^\n\r]*) _ {return makeNode("comment", {value: p.join('').trim()})}

decimal_point
  = "."

digit1_9
  = [1-9]
  
frac
  = decimal_point DIGIT+

int
  = zero / (digit1_9 DIGIT*)

minus
  = "-"

plus
  = "+"

zero
  = "0"
  
_ "whitespace"
  = [ \t\n\r]*
  
DIGIT  = [0-9]
