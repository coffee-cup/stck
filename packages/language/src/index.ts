export * from "./parser";
export * from "./interpret";
export * from "./types";
export * from "./stack";
