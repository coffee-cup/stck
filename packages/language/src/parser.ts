import { parse as parseRaw } from "./parserRaw";
import { Program } from "./types";

export const parse = (input: string): Program => parseRaw(input);
