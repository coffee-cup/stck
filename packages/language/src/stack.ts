import { Value } from "./types";

export type Stack = Value[];

export const createStack = (): Stack => [];

export const peek = (stack?: Stack): Value => {
  if (stack != null) {
    return stack[stack.length - 1];
  }

  return 0;
};

export const push = (value: Value, stack?: Stack): [Value, Stack] => {
  if (stack != null) {
    stack.push(value);
    return [value, stack];
  }

  const newStack = createStack();
  return push(value, newStack);
};

export const pop = (stack?: Stack): [Value, Stack?] => {
  if (stack != null) {
    const result = stack.pop();
    return [result ?? 0, stack];
  }

  return [0, stack];
};
