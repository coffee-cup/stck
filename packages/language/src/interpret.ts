import { repeat } from "lodash";
import { EvalError, typeError } from "./errors";
import {
  Expression,
  Identifier,
  Literal,
  Node,
  Call,
  NodeType,
  OP,
  Operator,
  Program,
  PushPop,
  Value,
  Loop,
  FunctionExpr,
} from "./types";

export type Stack = Value[];

export interface State {
  stacks: { [id: string]: Stack };
  ftable: { [name: string]: FunctionExpr };
}

export type VisitNode<T = Node> = (node: T, state: State) => void;

type Visitor = {
  [Type in Node["type"]]?: VisitNode<Extract<Node, { type: Type }>>;
};

const createEmptyState = (): State => ({
  stacks: {},
  ftable: {},
});

const isStackEmpty = (id: string, { stacks }: State): boolean =>
  stacks[id] == null || stacks[id].length === 0;

const peek = (id: string, { stacks }: State): Value => {
  const stack = stacks[id];
  if (stack == null) {
    return 0;
  }

  return stack[stack.length - 1];
};

const pop = (id: string, { stacks }: State): Value => {
  const stack = stacks[id];
  if (stack == null || stack.length === 0) {
    return 0;
  }

  return stack.pop()!;
};

const push = (value: Value, id: string, { stacks }: State) => {
  let stack = stacks[id];
  if (stack == null) {
    stack = [];
  }

  stack.push(value);
  stacks[id] = stack;
};

const isNumber = (v: Value): boolean => typeof v === "number";
const isString = (v: Value): boolean => typeof v === "string";

const visitProgram: VisitNode<Program> = (program, state) => {
  program.body.forEach(node => visit(node, state));
};

const visitPushPop: VisitNode<PushPop> = (node, state) => {
  const { left, right } = node;
  const id = right?.value;

  if (left && left.type === "literal" && id != null) {
    // push literal onto right
    push(left.value, id, state);
  } else if (left && left.type === "identifier" && id != null) {
    // pop value from left and push onto right
    const value = pop(left.value, state);
    push(value, id, state);
  } else if (left && left.type === "identifier") {
    // pop from left
    pop(left.value, state);
  } else if (left && left.type === "literal") {
    throw new EvalError("Cannot pop from literal", node);
  } else if (left == null) {
    throw new EvalError("left side of pushpop cannot be empty", node);
  }
};

const visitOperator: VisitNode<Operator> = (node, state) => {
  const { op, left, right } = node;

  if (op === "?") {
    // single operand
    if (right != null) {
      throw new EvalError(
        `${op} operator requires only leftmost operand`,
        node,
      );
    }

    const val = peek(left.value, state);
    if (val === 0) {
      state.stacks[left.value] = [];
    }
  } else {
    // multiple operands
    const val1: any = pop(left.value, state);
    const val2: any =
      right == null
        ? pop(left.value, state)
        : right.type === "literal"
        ? right.value
        : pop(right.value, state);

    let result: Value = 0;

    if (op === "+") {
      result = val1 + val2;
    } else if (op === "-") {
      if (isString(val1)) {
        throw new EvalError(typeError("subtraction", "number", val1), left);
      } else if (isString(val2)) {
        throw new EvalError(typeError("subtraction", "number", val2), right);
      }

      result = val1 - val2;
    } else if (op === "*") {
      if (isString(val2)) {
        throw new EvalError(typeError("multiplication", "number", val2), right);
      }

      result =
        isNumber(val1) && isNumber(val2) ? val1 * val2 : repeat(val1, val2);
    } else if (op === "/") {
      if (isString(val1)) {
        throw new EvalError(typeError("division", "number", val1), left);
      } else if (isString(val2)) {
        throw new EvalError(typeError("division", "number", val2), right);
      }

      result = val1 / val2;
    } else {
      throw new EvalError(`Operator ${op} not recognized`, node);
    }

    push(result, left.value, state);
  }
};

const visitFunction: VisitNode<FunctionExpr> = (node, state) => {
  node.body.forEach(node => visit(node, state));
};

const visitCall: VisitNode<Call> = (node, state) => {
  const fn = state.ftable[node.name];

  if (fn == null) {
    throw new EvalError(`Function ${node.name} is not defined`, node);
  }

  visitFunction(fn, state);
};

const visitLoop: VisitNode<Loop> = (node, state) => {
  while (!isStackEmpty(node.stack, state)) {
    node.body.forEach(n => visit(n, state));
  }
};

const visitors: Visitor = {
  program: visitProgram,
  pushpop: visitPushPop,
  operator: visitOperator,
  call: visitCall,
  loop: visitLoop,
};

const visit: VisitNode<Node> = (node: Node, state: State) => {
  const fn = visitors[node.type] as VisitNode<Node>;

  if (fn != null) {
    fn(node as any, state);
  }
};

const createFTable = (program: Program, state: State) => {
  program.body.forEach(expr => {
    if (expr.type === "function") {
      if (state.ftable[expr.name] != null) {
        throw new EvalError(`Function ${expr.name} cannot be redefined`, expr);
      }

      state.ftable[expr.name] = expr;
    }
  });
};

export const interpret = (program: Program): State => {
  const state = createEmptyState();

  createFTable(program, state);

  visit(program, state);
  return state;
};
