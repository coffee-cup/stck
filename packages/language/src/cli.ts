import * as commander from "commander";
import * as fs from "fs";
import * as getStdin from "get-stdin";
import * as path from "path";
import { interpret } from "./interpret";
import { parse } from "./parser";
import { Value } from "./types";

const packageJson = JSON.parse(
  fs.readFileSync(path.resolve(__dirname, "../package.json"), "utf8"),
);
const version: string = packageJson.version;

const readFile = (filename: string): Promise<string> => {
  const fullpath = path.resolve(process.cwd(), filename);
  return new Promise((resolve, reject) => {
    fs.readFile(fullpath, "utf8", (err, data) => {
      if (err != null) {
        reject(err);
      } else {
        resolve(data);
      }
    });
  });
};

const log = (s: string) => {
  // tslint:disable-next-line:no-console
  console.log(s);
};

const run = async (input: Value[]) => {
  const program = new commander.Command();
  program.version(version);

  program.parse(process.argv);

  try {
    if (program.args.length !== 1) {
      throw new Error(
        program.args.length === 0 ? "Filename required" : "Too many args",
      );
    }

    const filename = program.args[0];
    const contents = await readFile(filename);
    const ast = parse(contents);
    const state = interpret(ast, input);

    const output = state.stacks.o || [];

    log(output.join(""));
  } catch (e) {
    if (e.location != null) {
      const loc = e.location.start;
      log(`${loc.line}:${loc.column} ${e.message}`);
    } else {
      log(e.message);
    }
  }
};

const numRegex = /^-?\d+\.?\d*$/;
export const parseStdin = (stdin: string): Value[] => {
  if (stdin == null || stdin === "") {
    return [];
  }

  return stdin.split("\n").map(line => {
    try {
      if (numRegex.test(line)) {
        return parseFloat(line);
      }
    } catch (e) {
      // do nothing
    }

    return line;
  });
};

getStdin().then(stdin => {
  const inputStack = parseStdin(stdin.trim());
  run(inputStack);
});
