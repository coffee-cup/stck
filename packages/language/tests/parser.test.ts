import { parse } from "../src";
import {
  Comment,
  FunctionExpr,
  Program,
  Node,
  Call,
  Loop,
  PushPop,
} from "../src/types";

const firstNode = (program: Program): Node => program.body[0];

describe("parser", () => {
  it("full", () => {
    const program = parse(`
# comment

a+a
a-2
a*c
hello/world

1>a
a>
a>b

(a
  a>b
)

{:foo
  a+b
}

:foo
`);

    expect(program).toMatchSnapshot();
  });

  it("call", () => {
    const ident = firstNode(parse(":foo")) as Call;
    expect(ident.type).toBe("call");
    expect(ident.name).toBe("foo");
    expect(ident).toMatchSnapshot();

    try {
      parse("1_$%invalid");
      expect(true).toBe(false);
    } catch (e) {}
  });

  it("operators", () => {
    expect(firstNode(parse("a+b"))).toMatchSnapshot();
    expect(firstNode(parse("a+2"))).toMatchSnapshot();

    expect(firstNode(parse("a?"))).toMatchSnapshot();
  });

  it("pushpop", () => {
    {
      const pushpop = firstNode(parse("a>b")) as PushPop;
      expect(pushpop.type).toBe("pushpop");

      expect(pushpop.left!.type).toBe("identifier");
      expect(pushpop.left!.value).toBe("a");

      expect(pushpop.right!.type).toBe("identifier");
      expect(pushpop.right!.value).toBe("b");
    }

    {
      const pushpop = firstNode(parse("1>b")) as PushPop;
      expect(pushpop.type).toBe("pushpop");

      expect(pushpop.left!.type).toBe("literal");
      expect(pushpop.left!.value).toBe(1);

      expect(pushpop.right!.type).toBe("identifier");
      expect(pushpop.right!.value).toBe("b");
    }

    {
      const pushpop = firstNode(parse("a>")) as PushPop;
      expect(pushpop.type).toBe("pushpop");

      expect(pushpop.left!.type).toBe("identifier");
      expect(pushpop.left!.value).toBe("a");

      expect(pushpop.right).toBe(null);
    }
  });

  it("loop", () => {
    {
      const loop = firstNode(
        parse(`
(a
a+b
)
`),
      ) as Loop;

      expect(loop.type).toBe("loop");
      expect(loop.stack).toBe("a");
      expect(loop).toMatchSnapshot();
    }

    {
      const loop = firstNode(parse(`(a)`)) as Loop;

      expect(loop).toMatchSnapshot();
    }
  });

  it("function", () => {
    {
      const func = firstNode(
        parse(`
{:foo
  a+b
}
`),
      ) as FunctionExpr;

      expect(func.type).toBe("function");
      expect(func.name).toBe("foo");
      expect(func).toMatchSnapshot();
    }

    {
      const func = firstNode(parse(`{:foo}`)) as FunctionExpr;
      expect(func).toMatchSnapshot();
    }
  });

  it("comments", () => {
    {
      const comment = firstNode(
        parse(`
  # this is a comment
  `),
      ) as Comment;

      expect(comment.type).toBe("comment");
      expect(comment.value).toBe("this is a comment");
      expect(comment).toMatchSnapshot;
    }

    {
      const comment = parse(`
  # this is a comment

  a+a

  # second comment
  `);

      expect(comment).toMatchSnapshot;
    }
  });
});
