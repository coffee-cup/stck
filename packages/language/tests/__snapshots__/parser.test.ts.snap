// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser call 1`] = `
Object {
  "end": Object {
    "column": 5,
    "line": 1,
    "offset": 4,
  },
  "name": "foo",
  "start": Object {
    "column": 1,
    "line": 1,
    "offset": 0,
  },
  "type": "call",
}
`;

exports[`parser full 1`] = `
Object {
  "body": Array [
    Object {
      "end": Object {
        "column": 1,
        "line": 4,
        "offset": 12,
      },
      "start": Object {
        "column": 1,
        "line": 2,
        "offset": 1,
      },
      "type": "comment",
      "value": "comment",
    },
    Object {
      "end": Object {
        "column": 4,
        "line": 4,
        "offset": 15,
      },
      "left": Object {
        "end": Object {
          "column": 2,
          "line": 4,
          "offset": 13,
        },
        "start": Object {
          "column": 1,
          "line": 4,
          "offset": 12,
        },
        "type": "identifier",
        "value": "a",
      },
      "op": "+",
      "right": Object {
        "end": Object {
          "column": 4,
          "line": 4,
          "offset": 15,
        },
        "start": Object {
          "column": 3,
          "line": 4,
          "offset": 14,
        },
        "type": "identifier",
        "value": "a",
      },
      "start": Object {
        "column": 1,
        "line": 4,
        "offset": 12,
      },
      "type": "operator",
    },
    Object {
      "end": Object {
        "column": 4,
        "line": 5,
        "offset": 19,
      },
      "left": Object {
        "end": Object {
          "column": 2,
          "line": 5,
          "offset": 17,
        },
        "start": Object {
          "column": 1,
          "line": 5,
          "offset": 16,
        },
        "type": "identifier",
        "value": "a",
      },
      "op": "-",
      "right": Object {
        "end": Object {
          "column": 4,
          "line": 5,
          "offset": 19,
        },
        "start": Object {
          "column": 3,
          "line": 5,
          "offset": 18,
        },
        "type": "literal",
        "value": 2,
      },
      "start": Object {
        "column": 1,
        "line": 5,
        "offset": 16,
      },
      "type": "operator",
    },
    Object {
      "end": Object {
        "column": 4,
        "line": 6,
        "offset": 23,
      },
      "left": Object {
        "end": Object {
          "column": 2,
          "line": 6,
          "offset": 21,
        },
        "start": Object {
          "column": 1,
          "line": 6,
          "offset": 20,
        },
        "type": "identifier",
        "value": "a",
      },
      "op": "*",
      "right": Object {
        "end": Object {
          "column": 4,
          "line": 6,
          "offset": 23,
        },
        "start": Object {
          "column": 3,
          "line": 6,
          "offset": 22,
        },
        "type": "identifier",
        "value": "c",
      },
      "start": Object {
        "column": 1,
        "line": 6,
        "offset": 20,
      },
      "type": "operator",
    },
    Object {
      "end": Object {
        "column": 12,
        "line": 7,
        "offset": 35,
      },
      "left": Object {
        "end": Object {
          "column": 6,
          "line": 7,
          "offset": 29,
        },
        "start": Object {
          "column": 1,
          "line": 7,
          "offset": 24,
        },
        "type": "identifier",
        "value": "hello",
      },
      "op": "/",
      "right": Object {
        "end": Object {
          "column": 12,
          "line": 7,
          "offset": 35,
        },
        "start": Object {
          "column": 7,
          "line": 7,
          "offset": 30,
        },
        "type": "identifier",
        "value": "world",
      },
      "start": Object {
        "column": 1,
        "line": 7,
        "offset": 24,
      },
      "type": "operator",
    },
    Object {
      "end": Object {
        "column": 4,
        "line": 9,
        "offset": 40,
      },
      "left": Object {
        "end": Object {
          "column": 2,
          "line": 9,
          "offset": 38,
        },
        "start": Object {
          "column": 1,
          "line": 9,
          "offset": 37,
        },
        "type": "literal",
        "value": 1,
      },
      "right": Object {
        "end": Object {
          "column": 4,
          "line": 9,
          "offset": 40,
        },
        "start": Object {
          "column": 3,
          "line": 9,
          "offset": 39,
        },
        "type": "identifier",
        "value": "a",
      },
      "start": Object {
        "column": 1,
        "line": 9,
        "offset": 37,
      },
      "type": "pushpop",
    },
    Object {
      "end": Object {
        "column": 3,
        "line": 10,
        "offset": 43,
      },
      "left": Object {
        "end": Object {
          "column": 2,
          "line": 10,
          "offset": 42,
        },
        "start": Object {
          "column": 1,
          "line": 10,
          "offset": 41,
        },
        "type": "identifier",
        "value": "a",
      },
      "right": null,
      "start": Object {
        "column": 1,
        "line": 10,
        "offset": 41,
      },
      "type": "pushpop",
    },
    Object {
      "end": Object {
        "column": 4,
        "line": 11,
        "offset": 47,
      },
      "left": Object {
        "end": Object {
          "column": 2,
          "line": 11,
          "offset": 45,
        },
        "start": Object {
          "column": 1,
          "line": 11,
          "offset": 44,
        },
        "type": "identifier",
        "value": "a",
      },
      "right": Object {
        "end": Object {
          "column": 4,
          "line": 11,
          "offset": 47,
        },
        "start": Object {
          "column": 3,
          "line": 11,
          "offset": 46,
        },
        "type": "identifier",
        "value": "b",
      },
      "start": Object {
        "column": 1,
        "line": 11,
        "offset": 44,
      },
      "type": "pushpop",
    },
    Object {
      "body": Array [
        Object {
          "end": Object {
            "column": 6,
            "line": 14,
            "offset": 57,
          },
          "left": Object {
            "end": Object {
              "column": 4,
              "line": 14,
              "offset": 55,
            },
            "start": Object {
              "column": 3,
              "line": 14,
              "offset": 54,
            },
            "type": "identifier",
            "value": "a",
          },
          "right": Object {
            "end": Object {
              "column": 6,
              "line": 14,
              "offset": 57,
            },
            "start": Object {
              "column": 5,
              "line": 14,
              "offset": 56,
            },
            "type": "identifier",
            "value": "b",
          },
          "start": Object {
            "column": 3,
            "line": 14,
            "offset": 54,
          },
          "type": "pushpop",
        },
      ],
      "end": Object {
        "column": 2,
        "line": 15,
        "offset": 59,
      },
      "stack": "a",
      "start": Object {
        "column": 1,
        "line": 13,
        "offset": 49,
      },
      "type": "loop",
    },
    Object {
      "body": Array [
        Object {
          "end": Object {
            "column": 6,
            "line": 18,
            "offset": 72,
          },
          "left": Object {
            "end": Object {
              "column": 4,
              "line": 18,
              "offset": 70,
            },
            "start": Object {
              "column": 3,
              "line": 18,
              "offset": 69,
            },
            "type": "identifier",
            "value": "a",
          },
          "op": "+",
          "right": Object {
            "end": Object {
              "column": 6,
              "line": 18,
              "offset": 72,
            },
            "start": Object {
              "column": 5,
              "line": 18,
              "offset": 71,
            },
            "type": "identifier",
            "value": "b",
          },
          "start": Object {
            "column": 3,
            "line": 18,
            "offset": 69,
          },
          "type": "operator",
        },
      ],
      "end": Object {
        "column": 2,
        "line": 19,
        "offset": 74,
      },
      "name": "foo",
      "start": Object {
        "column": 1,
        "line": 17,
        "offset": 61,
      },
      "type": "function",
    },
    Object {
      "end": Object {
        "column": 5,
        "line": 21,
        "offset": 80,
      },
      "name": "foo",
      "start": Object {
        "column": 1,
        "line": 21,
        "offset": 76,
      },
      "type": "call",
    },
  ],
  "end": Object {
    "column": 1,
    "line": 22,
    "offset": 81,
  },
  "start": Object {
    "column": 1,
    "line": 1,
    "offset": 0,
  },
  "type": "program",
}
`;

exports[`parser function 1`] = `
Object {
  "body": Array [
    Object {
      "end": Object {
        "column": 6,
        "line": 3,
        "offset": 12,
      },
      "left": Object {
        "end": Object {
          "column": 4,
          "line": 3,
          "offset": 10,
        },
        "start": Object {
          "column": 3,
          "line": 3,
          "offset": 9,
        },
        "type": "identifier",
        "value": "a",
      },
      "op": "+",
      "right": Object {
        "end": Object {
          "column": 6,
          "line": 3,
          "offset": 12,
        },
        "start": Object {
          "column": 5,
          "line": 3,
          "offset": 11,
        },
        "type": "identifier",
        "value": "b",
      },
      "start": Object {
        "column": 3,
        "line": 3,
        "offset": 9,
      },
      "type": "operator",
    },
  ],
  "end": Object {
    "column": 2,
    "line": 4,
    "offset": 14,
  },
  "name": "foo",
  "start": Object {
    "column": 1,
    "line": 2,
    "offset": 1,
  },
  "type": "function",
}
`;

exports[`parser function 2`] = `
Object {
  "body": Array [],
  "end": Object {
    "column": 7,
    "line": 1,
    "offset": 6,
  },
  "name": "foo",
  "start": Object {
    "column": 1,
    "line": 1,
    "offset": 0,
  },
  "type": "function",
}
`;

exports[`parser loop 1`] = `
Object {
  "body": Array [
    Object {
      "end": Object {
        "column": 4,
        "line": 3,
        "offset": 7,
      },
      "left": Object {
        "end": Object {
          "column": 2,
          "line": 3,
          "offset": 5,
        },
        "start": Object {
          "column": 1,
          "line": 3,
          "offset": 4,
        },
        "type": "identifier",
        "value": "a",
      },
      "op": "+",
      "right": Object {
        "end": Object {
          "column": 4,
          "line": 3,
          "offset": 7,
        },
        "start": Object {
          "column": 3,
          "line": 3,
          "offset": 6,
        },
        "type": "identifier",
        "value": "b",
      },
      "start": Object {
        "column": 1,
        "line": 3,
        "offset": 4,
      },
      "type": "operator",
    },
  ],
  "end": Object {
    "column": 2,
    "line": 4,
    "offset": 9,
  },
  "stack": "a",
  "start": Object {
    "column": 1,
    "line": 2,
    "offset": 1,
  },
  "type": "loop",
}
`;

exports[`parser loop 2`] = `
Object {
  "body": Array [],
  "end": Object {
    "column": 4,
    "line": 1,
    "offset": 3,
  },
  "stack": "a",
  "start": Object {
    "column": 1,
    "line": 1,
    "offset": 0,
  },
  "type": "loop",
}
`;

exports[`parser operators 1`] = `
Object {
  "end": Object {
    "column": 4,
    "line": 1,
    "offset": 3,
  },
  "left": Object {
    "end": Object {
      "column": 2,
      "line": 1,
      "offset": 1,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
    "type": "identifier",
    "value": "a",
  },
  "op": "+",
  "right": Object {
    "end": Object {
      "column": 4,
      "line": 1,
      "offset": 3,
    },
    "start": Object {
      "column": 3,
      "line": 1,
      "offset": 2,
    },
    "type": "identifier",
    "value": "b",
  },
  "start": Object {
    "column": 1,
    "line": 1,
    "offset": 0,
  },
  "type": "operator",
}
`;

exports[`parser operators 2`] = `
Object {
  "end": Object {
    "column": 4,
    "line": 1,
    "offset": 3,
  },
  "left": Object {
    "end": Object {
      "column": 2,
      "line": 1,
      "offset": 1,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
    "type": "identifier",
    "value": "a",
  },
  "op": "+",
  "right": Object {
    "end": Object {
      "column": 4,
      "line": 1,
      "offset": 3,
    },
    "start": Object {
      "column": 3,
      "line": 1,
      "offset": 2,
    },
    "type": "literal",
    "value": 2,
  },
  "start": Object {
    "column": 1,
    "line": 1,
    "offset": 0,
  },
  "type": "operator",
}
`;

exports[`parser operators 3`] = `
Object {
  "end": Object {
    "column": 3,
    "line": 1,
    "offset": 2,
  },
  "left": Object {
    "end": Object {
      "column": 2,
      "line": 1,
      "offset": 1,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
    "type": "identifier",
    "value": "a",
  },
  "op": "?",
  "right": null,
  "start": Object {
    "column": 1,
    "line": 1,
    "offset": 0,
  },
  "type": "operator",
}
`;
